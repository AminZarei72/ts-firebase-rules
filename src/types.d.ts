
/* ========================================== */
/* 
//Note:this is buggy on ts watch
 export type PackageJson = typeof import('../package.json')
  */
export type PackageJson = { version: string }
/* ========================================== */
