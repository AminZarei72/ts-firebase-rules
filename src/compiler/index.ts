import * as fs from 'fs-extra';
import { concatAllNativeFns } from './concatAllNativeFns';
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
// const sourceDir = '../'
// const compiledTsFile = sourceDir + 'build/typedFns/groups/create_groups.js'
/* const compiledTsFile = sourceDir + '/build/finalFile.js' */ /* enableThis */
// const compiledTsFile = sourceDir + '/build/finalFile.js'
// export const native = sourceDir + 'src/nativeFunctions/'
// const rule = sourceDir + 'fbFiles/firestore.rules'
/* const outputFilePath = sourceDir + 'build/firestore.rules' */
// const outputFilePath = '_/firestore.rules'
// const compiledTsReplacer = '//$compiledFunctionsWillBeReplacedWithThis'
// const nativeFunctionsTextToReplace = '//$nativeFirestoreFunctionsWillBeReplacedWithThis'
/* --------------- */
export async function build(args: {
    compiledTsPath: string,
    targetDotRulesPath: string,
    nativeFilesPath: string,
    nativeFunctionsTextToReplace: string,
    outputFilePath: string,
    additionalNativeHelperFnsPath: string,
}) {
    /* --------------- */
    console.log('compiling')
    /* --------------- */
    /* todo:we can get this from user */
    let finalRules = fs.readFileSync(args.targetDotRulesPath, { encoding: 'utf-8' })
    // let finalRules = rules
    /* --------------- */
    // finalRules = finalRules.replace(nativeFunctionsTextToReplace, concatAllNativeFns())
    /* --------------- */
    // finalRules=converters1(compiledTs)
    const compiledTs = fs.readFileSync(args.compiledTsPath, { encoding: 'utf-8' })
    // const compiledTs = newFile ? newFile : fs.readFileSync(compiledTsFile, { encoding: 'utf-8' })
    // const compiledTs = newFile/*  ? newFile : fs.readFileSync(compiledTsFile, { encoding: 'utf-8' }) */
    // finalRules = finalRules.replace(nativeFunctionsTextToReplace, converters1(compiledTs))
    /* --------------- */
    // console.log(args.nativeFilesPath)

    const finalF = `
    //=======================================================
    ${concatAllNativeFns(args.nativeFilesPath, args.additionalNativeHelperFnsPath)}
    //=======================================================
    CompileTheCompiledTsHere
    //=======================================================
    `/* todo:CompileTheCompiledTsHere */

    finalRules = finalRules.replace(args.nativeFunctionsTextToReplace, finalF)

    // finalRules = finalRules.replace(compiledTsReplacer, compiledTs)
    /* --------------- */
    fs.ensureFileSync(args.outputFilePath)
    console.log(args.outputFilePath, finalRules)
    fs.writeFileSync(args.outputFilePath, finalRules)
    /* --------------- */
    /* target.split(search).join(replacement); */
    /* =============================== */
    // function escapeRegExp(string) {
    //     return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
    // }

    // function replaceAll(str, find, replace) {
    //     return str.replace(new RegExp(escapeRegExp(find), 'g'), replace);
    // }
    /* =============================== */
    return null
} 