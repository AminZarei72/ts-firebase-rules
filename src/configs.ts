export const appName = 'ts-firebase-rules'
export const appName_cli = 'tsfr' 
export const fakeDelay = 'development' === process.env.NODE_ENV ? 0 : 500

export const tsfr_config_fileName = `tsfr.config.json`


export const additionalHelperFnsPath = `../../otherRequiredFiles/helpers/index.ts`
export const tsconfig_forCompilingPath = `../../otherRequiredFiles/tsconfig_forCompiling.json`
export const additionalHelperTypesPath = `../../otherRequiredFiles/tsfrTypes/index.ts`
export const tscPath = `../../otherRequiredFiles/typescript/bin/tsc`
export const compiledTsPath = `../../otherRequiredFiles/compiledTs.js`

export const additionalNativeHelperFnsPath = `../../otherRequiredFiles/helpers/nativeFunctions.rules`

export const nativeFunctionsTextToReplace = '//<<$nativeFirestoreFunctionsWillBeReplacedWithThis>>;'
