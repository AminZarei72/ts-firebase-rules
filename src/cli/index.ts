#!/usr/bin/env node
import sade from 'sade';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import { addCreateToCli } from './addCreateToCli';
const pkg = require('../../package.json');
import * as configs from '../configs'
/* ==================================== */
async function initCli() {
    console.log('***' + chalk.bgBlue('TS') + (chalk.red('FR')) + '***');
    console.info(`***wellcome to ${configs.appName} cli***\n`)
    const cli = sade(configs.appName_cli);
    const templates = { todoList: '  ' }
    const userCurrentPath = (await fs.realpath(process.cwd()))
    addCreateToCli({ templates, userCurrentPath, pkg })
    cli.parse(process.argv);
}
/* ==================================== */
initCli()


