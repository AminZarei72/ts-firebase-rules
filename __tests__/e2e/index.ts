import shell from 'shelljs'
import * as fs from 'fs-extra'
import * as configs from '../../src/configs'
/* ======================================== */
test('"create" works', async () => {
    expect(1 + 2).toBe(3)
    const res = shell.exec('npx tsfr')
    expect(res.stdout.includes(configs.appName)).toBeTruthy()

    // const res2 = shell.exec('npx tsfr create')
    // const res2 = shell.exec('npx tsfr create --template todoList')
    // expect(res2.stdout.includes('creating')).toBeTruthy()

    const currentDir = `./tmp/e2e1Files`
    const userChoosedDir = `tsfrPrj`
    /* ----------- */
    shell.rm('-rf', currentDir)
    // fs.mkdirSync(currentDir)
    fs.ensureDirSync(currentDir)

    await testCreate({ currentDir: `${currentDir}/prj1`, userChoosedDir })



})
/* ======================================== */
async function testCreate(args: { currentDir: string, userChoosedDir: string }) {
    console.log('user is initilizing its tsfr project...')
    fs.mkdirSync(args.currentDir)
    const { code } = shell.exec(`cd ${args.currentDir} && npx tsfr create ${args.userChoosedDir} --template todoList`)
    if (code != 0) throw 'failed'
    /* these will be made from scratch */
    expect(fs.existsSync(`${args.currentDir}/tsfr.config.json`)).toBeTruthy()
    /* these will get coppied from template folder */
    // expect(fs.existsSync(`${args.currentDir}/${args.userChoosedDir}/index.ts`)).toBeTruthy()

    expect(fs.readdirSync(`${args.currentDir}/${args.userChoosedDir}/`, {
        withFileTypes: true
    })).toMatchSnapshot('filesList')

}
/* ======================================== */
