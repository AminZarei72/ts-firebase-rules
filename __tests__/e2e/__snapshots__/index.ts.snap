// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"create" works: filesList 1`] = `
Array [
  Dirent {
    "name": "build",
    Symbol(type): 2,
  },
  Dirent {
    "name": "functions",
    Symbol(type): 2,
  },
  Dirent {
    "name": "index.ts",
    Symbol(type): 1,
  },
  Dirent {
    "name": "native",
    Symbol(type): 2,
  },
  Dirent {
    "name": "tsconfig.json",
    Symbol(type): 1,
  },
]
`;

exports[`"create" works: finalBuild 1`] = `
Object {
  "data": Array [
    10,
    32,
    32,
    32,
    32,
    47,
    47,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    10,
    32,
    32,
    32,
    32,
    110,
    97,
    116,
    105,
    118,
    101,
    70,
    117,
    110,
    99,
    116,
    105,
    111,
    110,
    115,
    32,
    10,
    32,
    10,
    10,
    32,
    32,
    32,
    32,
    47,
    47,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    10,
    32,
    32,
    32,
    32,
    32,
    32,
    32,
    32,
    34,
    117,
    115,
    101,
    32,
    115,
    116,
    114,
    105,
    99,
    116,
    34,
    59,
    10,
    32,
    32,
    32,
    32,
    79,
    98,
    106,
    101,
    99,
    116,
    46,
    100,
    101,
    102,
    105,
    110,
    101,
    80,
    114,
    111,
    112,
    101,
    114,
    116,
    121,
    40,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    44,
    32,
    34,
    95,
    95,
    101,
    115,
    77,
    111,
    100,
    117,
    108,
    101,
    34,
    44,
    32,
    123,
    32,
    118,
    97,
    108,
    117,
    101,
    58,
    32,
    116,
    114,
    117,
    101,
    32,
    125,
    41,
    59,
    10,
    32,
    32,
    32,
    32,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    46,
    99,
    114,
    101,
    97,
    116,
    101,
    95,
    116,
    111,
    100,
    111,
    32,
    61,
    32,
    118,
    111,
    105,
    100,
    32,
    48,
    59,
    10,
    32,
    32,
    32,
    32,
    102,
    117,
    110,
    99,
    116,
    105,
    111,
    110,
    32,
    99,
    114,
    101,
    97,
    116,
    101,
    95,
    116,
    111,
    100,
    111,
    40,
    105,
    100,
    41,
    32,
    123,
    10,
    32,
    32,
    32,
    32,
    32,
    32,
    32,
    32,
    114,
    101,
    116,
    117,
    114,
    110,
    32,
    116,
    114,
    117,
    101,
    59,
    10,
    32,
    32,
    32,
    32,
    125,
    10,
    32,
    32,
    32,
    32,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    46,
    99,
    114,
    101,
    97,
    116,
    101,
    95,
    116,
    111,
    100,
    111,
    32,
    61,
    32,
    99,
    114,
    101,
    97,
    116,
    101,
    95,
    116,
    111,
    100,
    111,
    59,
    32,
    32,
    32,
    32,
    34,
    117,
    115,
    101,
    32,
    115,
    116,
    114,
    105,
    99,
    116,
    34,
    59,
    10,
    32,
    32,
    32,
    32,
    79,
    98,
    106,
    101,
    99,
    116,
    46,
    100,
    101,
    102,
    105,
    110,
    101,
    80,
    114,
    111,
    112,
    101,
    114,
    116,
    121,
    40,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    44,
    32,
    34,
    95,
    95,
    101,
    115,
    77,
    111,
    100,
    117,
    108,
    101,
    34,
    44,
    32,
    123,
    32,
    118,
    97,
    108,
    117,
    101,
    58,
    32,
    116,
    114,
    117,
    101,
    32,
    125,
    41,
    59,
    10,
    32,
    32,
    32,
    32,
    95,
    95,
    101,
    120,
    112,
    111,
    114,
    116,
    83,
    116,
    97,
    114,
    40,
    99,
    114,
    101,
    97,
    116,
    101,
    95,
    116,
    111,
    100,
    111,
    95,
    49,
    44,
    32,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    41,
    59,
    32,
    32,
    32,
    32,
    34,
    117,
    115,
    101,
    32,
    115,
    116,
    114,
    105,
    99,
    116,
    34,
    59,
    10,
    32,
    32,
    32,
    32,
    79,
    98,
    106,
    101,
    99,
    116,
    46,
    100,
    101,
    102,
    105,
    110,
    101,
    80,
    114,
    111,
    112,
    101,
    114,
    116,
    121,
    40,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    44,
    32,
    34,
    95,
    95,
    101,
    115,
    77,
    111,
    100,
    117,
    108,
    101,
    34,
    44,
    32,
    123,
    32,
    118,
    97,
    108,
    117,
    101,
    58,
    32,
    116,
    114,
    117,
    101,
    32,
    125,
    41,
    59,
    10,
    32,
    32,
    32,
    32,
    95,
    95,
    101,
    120,
    112,
    111,
    114,
    116,
    83,
    116,
    97,
    114,
    40,
    116,
    111,
    100,
    111,
    115,
    95,
    49,
    44,
    32,
    101,
    120,
    112,
    111,
    114,
    116,
    115,
    41,
    59,
    10,
    32,
    32,
    32,
    32,
    47,
    47,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    61,
    10,
    32,
    32,
    32,
    32,
    32,
    10,
    32,
    10,
  ],
  "type": "Buffer",
}
`;
